catalog/dags/popularity/sql.py

```diff
@@ -0,0 +1,362 @@
+        postgres_conn_id=postgres_conn_id, default_statement_timeout=10.0
+    )
+    popularity_metrics_columns_string = ",\n            ".join(
+        f"{c.name} {c.definition}" for c in POPULARITY_METRICS_TABLE_COLUMNS
+    )
+    query = dedent(
+        f"""
+        CREATE TABLE public.{sql_info.metrics_table} (
+          {popularity_metrics_columns_string}
+        );
+        """
+    )
+    postgres.run(query)
+
+
+@task
+@setup_sql_info_for_media_type
+def update_media_popularity_metrics(
+    postgres_conn_id: str,
+    media_type: str,
+    popularity_metrics: dict,
+    sql_info: SQLInfo = None,
+    task: AbstractOperator = None,
+):
+    postgres = PostgresHook(
+        postgres_conn_id=postgres_conn_id,
+        default_statement_timeout=PostgresHook.get_execution_timeout(task),
+    )
+
+    column_names = [c.name for c in POPULARITY_METRICS_TABLE_COLUMNS]
+
+    # Note that we do not update the val and constant. That is only done during the
+    # calculation tasks. In other words, we never want to clear out the current value of
+    # the popularity constant unless we're already done calculating the new one, since
+    # that can be a time consuming process.
+    updates_string = ",\n          ".join(
+        f"{c}=EXCLUDED.{c}"
+        for c in column_names
+        if c not in [PARTITION, CONSTANT, VALUE]
+    )
+    popularity_metric_inserts = _get_popularity_metric_insert_values_string(
+        popularity_metrics
+    )
+
+    query = dedent(
+        f"""
+        INSERT INTO public.{sql_info.metrics_table} (
+          {', '.join(column_names)}
+        ) VALUES
+          {popularity_metric_inserts}
+        ON CONFLICT ({PARTITION})
+        DO UPDATE SET
+          {updates_string}
+        ;
+        """
+    )
+    return postgres.run(query)
+
+
+@task
+@setup_sql_info_for_media_type
+def calculate_media_popularity_percentile_value(
+    postgres_conn_id: str,
+    provider: str,
+    media_type: str,
+    sql_info: SQLInfo = None,
+    task: AbstractOperator = None,
+):
+    postgres = PostgresHook(
+        postgres_conn_id=postgres_conn_id,
+        default_statement_timeout=PostgresHook.get_execution_timeout(task),
+    )
+
+    # Calculate the percentile value. E.g. if `percentile` = 0.80, then we'll
+    # calculate the _value_ of the 80th percentile for this provider's
+    # popularity metric.
+    calculate_new_percentile_value_query = dedent(
+        f"""
+        SELECT {sql_info.popularity_percentile_fn}({PARTITION}, {METRIC}, {PERCENTILE})
+        FROM {sql_info.metrics_table}
+        WHERE {col.PROVIDER.db_name}='{provider}';
+        """
+    )
+
+    return postgres.run(calculate_new_percentile_value_query, handler=single_value)
+
+
+@task
+@setup_sql_info_for_media_type
+def update_percentile_and_constants_values_for_provider(
+    postgres_conn_id: str,
+    provider: str,
+    raw_percentile_value: float,
+    media_type: str,
+    popularity_metrics: dict,
+    sql_info: SQLInfo = None,
```

Comment:
Is the `None` default required here? Won't it always be supplied by `@setup_sql_info_for_media_type`? Is the default to satisfy a typecheck for the caller, I guess so editors don't think the callsite _needs_ to pass that kwarg?

I guess if the decorator implemented [ParamSpec](https://docs.python.org/3/library/typing.html#typing.ParamSpec) on the `callable` type of the wrapped function and itself set the kwarg to optional it could handle the call site documentation. Complex though, so not worth it. It does a bit of confusion in the function implementation though, as `sql_info` would never _actually_ be `None`... the decorator itself overrides the value, even if the caller explicitly set it to `None`. I'd be tempted to check that `sql_info is not None` if I was working in this function and didn't have a nuanced understanding of how `setup_sql_info_for_media_type` worked.

To summarise: the `None` default is confusing when making changes to the function, if the decorator's implementation details and implications1 aren't well understood. The risk of error is low (checking `sql_info is not None` won't hurt, it's just technically useless) so it isn't a blocking issue, just a nit pick. Considering the complexity that it could take to solve this, if editors complain about not passing `sql_info`, it might not be worth it to make any changes here. On the other hand, if the default parameter isn't necessary to satisfy basic editor introspection, it would improve the clarity of these functions to remove the default `None` value.

Reply:
I agree on all counts -- unfortunately, the `None` default does appear to be required when the function is also using the `@task` decorator for TaskFlow, or we get errors at DAG parsing :(